src/prefect/runner/utils.py

```diff
@@ -6,76 +6,79 @@
 from prefect import __version__ as PREFECT_VERSION
 
 
-def _inject_schemas_into_generated_openapi(
-    webserver: FastAPI, schemas: Dict[str, Any]
+def inject_schemas_into_openapi(
+    webserver: FastAPI, deployment_schemas: Dict[str, Any]
 ) -> Dict[str, Any]:
-    """Injects the parameter schemas into the Runner's Webserver OpenAPI schema
+    """
+    Augments the webserver's OpenAPI schema with additional schemas from deployments.
 
     Args:
-        - webserver: the webserver to inject the schemas into
-        - schemas: the schemas to inject
+        webserver: The FastAPI instance representing the webserver.
+        deployment_schemas: A dictionary of deployment schemas to integrate.
+
+    Returns:
+        The augmented OpenAPI schema dictionary.
     """
     openapi_schema = get_openapi(
         title="FastAPI Prefect Runner", version=PREFECT_VERSION, routes=webserver.routes
     )
 
-    # Update the components/schemas with definitions from the deployment schemas
-    _merge_definitions_into_components(schemas, openapi_schema)
-
-    # Update the paths with correct references based on the components
-    _update_paths_with_correct_refs(openapi_schema)
+    merge_definitions(deployment_schemas, openapi_schema)
+    update_refs_to_components(openapi_schema)
 
     return openapi_schema
 
 
-def _merge_definitions_into_components(
-    input_schemas: Dict[str, Any], openapi_schema: Dict[str, Any]
+def merge_definitions(
+    deployment_schemas: Dict[str, Any], openapi_schema: Dict[str, Any]
 ) -> None:
-    """Merges the input schemas into the OpenAPI schema's components/schemas
+    """
+    Integrates definitions from deployment schemas into the OpenAPI components.
 
     Args:
-        - input_schemas: the schemas to merge into the OpenAPI schema
-        - openapi_schema: the OpenAPI schema to merge the schemas into
+        deployment_schemas: A dictionary of deployment-specific schemas.
+        openapi_schema: The base OpenAPI schema to update.
     """
-    components_schemas = openapi_schema.setdefault("components", {}).setdefault(
-        "schemas", {}
-    )
-    for schema in input_schemas.values():
-        if "definitions" in schema:
-            for def_name, def_value in schema["definitions"].items():
-                # Recursively update all $ref in the schema definition
-                _recursively_update_refs(def_value, "#/components/schemas/")
-                components_schemas[def_name] = def_value
+    components = openapi_schema.setdefault("components", {}).setdefault("schemas", {})
+    for definitions in deployment_schemas.values():
+        if "definitions" in definitions:
+            for def_name, def_schema in definitions["definitions"].items():
+                update_refs_in_schema(def_schema, "#/components/schemas/")
+                components[def_name] = def_schema
```

Comment:
I've had my share of nasty surprises from updating dictionaries in place. Since the end result of this chain is to pass the result back and set an attribute on `webserver`, I recommend updating these utils not to alter their inputs and return a new dictionary instead. That will make them more predictable to their callers.

In particular, I believe the definitions in `deployment_schemas` will be different after calling this function, which would surprise me as a caller.

Reply:
changed to copying them here in [221e6d1](https://github.com/PrefectHQ/prefect/pull/11333/commits/221e6d1acf2421ffd09e141fad7897e20da8bb00)